### PR TITLE
Extensible input module + new cursor

### PR DESCRIPTION
**Pull request by maxouellet**
These changes add the new HoloToolkit Input Module and the new cursor implementation. Documentation for the new input scripts has been updated, and all existing test scenes and prefabs have been fixed so that their behavior remains the same. This pull request is targeting the ExtensibleInputModule branch, so that we can stabilize in there as necessary and give people some time to take a look at the changes before they go to master.

The main changes are the following:

- New extensible input module, as discussed [here](https://github.com/Microsoft/HoloToolkit-Unity/issues/277). The old GesturesManager, HandsManager and GazeManager have been removed, since they are fully replaced by the new scripts. Existing scripts that dependend on those have been modified to use the new input module.

- New extensible cursor implementation, as described by @paseb [here](https://github.com/Microsoft/HoloToolkit-Unity/issues/277). This adds a new Cursor prefab (DefaultCursor). The existing cursors prefabs have been modified so that they use the new cursor scripts, and the old cursor scripts have been removed.

- Singleton has been modified to be more streamlined and predictable. Existing Singletons have been adjusted accordingly. The main breaking change here is that singleton classes that implement Awake or OnDestroy should override them, and make sure to call base.Awake() and base.OnDestroy().

### CODE REVIEW

**Comment by HodgsonSDAS** [comment]
@maxouellet `Cursor.fbx` still contains those project specific animations and geometry.

File: Assets/HoloToolkit/Input/Scripts/Gaze/BaseRayStabilizer.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+﻿using UnityEngine;
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Input/Scripts/Gaze/BaseRayStabilizer.cs]: Missing licencing header

File: Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs [hunk 2]
```diff
@@ -0,0 +1,345 @@
+﻿using System.Collections.Generic;
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs]: Missing licensing header

File: Assets/HoloToolkit/Input/Scripts/UI/Button.cs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        public void DehydrateButton()
+        {
+            if (ButtonAnimator != null && ButtonAnimator.isInitialized)
+            {
+                var animatorHashes = ButtonAnimator.parameters;
+                for (int i = 0; i < animatorHashes.Length; i++)
+                {
+                    if (animatorHashes[i].nameHash == deHydrateButtonId)
+                    {
+                        ButtonAnimator.SetTrigger(deHydrateButtonId);
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Input/Scripts/UI/Button.cs]: Does this trigger also call `gameObject.setActive(false)`?

File: Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+                return null;
+            }
+
+            // Return the minimum distance hit within the first layer that has hits.
+            // In other words, sort all hit objects first by layerMask, then by distance.
+            for (int layerMaskIdx = 0; layerMaskIdx < RaycastLayerMasks.Length; layerMaskIdx++)
+            {
+                RaycastHit? minHit = null;
+
+                foreach (RaycastHit h in hits)
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs]: could be a for loop because this is embedded in an update loop.

File: Assets/HoloToolkit/Input/Scripts/KeyboardManager.cs [hunk 5]
```diff
@@ -0,0 +1,237 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// TODO - Review this code before uploading to the HoloToolkit
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Input/Scripts/KeyboardManager.cs]: Fix licencing header

File: Assets/HoloToolkit/Input/Scripts/Gaze/BaseRayStabilizer.cs (hunk 1 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Gaze/BaseRayStabilizer.cs]: Fixed

File: Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs (hunk 2 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs]: Fixed

File: Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs (hunk 4 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs]: Fixed

File: Assets/HoloToolkit/Input/Scripts/KeyboardManager.cs (hunk 5 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/KeyboardManager.cs]: Fixed

File: Assets/HoloToolkit/Input/Scripts/UI/Button.cs (hunk 3 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/UI/Button.cs]: It does not, but I don't think it should in this case, as this button is . The dehydration animation itself sets the button's renderer to inactive. I modified the animation so that it also sets the button collider to inactive, which I think addresses the issue.

That being said, I'm not convinced this button script is ready for main HoloToolkit yet. As such, I renamed it to TestButton and moved it to the Tests/Scripts location, so that it serves as an example script for the test scene, rather than something that people should start building on. Some design work would be required to make this into a great reusable button script that works across the board.

**Comment by maxouellet** [comment]
@HodgsonSDAS Ahh I finally understand what you mean, there's animations embedded in that cursor FBX. Working on fixing that, I'll just remove the animations that are in the cursor and essentially regenerate the FBX.

File: Assets/HoloToolkit/Input/Scripts/UI/Button.cs (hunk 3 above)
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Input/Scripts/UI/Button.cs]: I think it's a good idea to let developers figure out how they want to handle buttons and just have an example.

File: Assets/HoloToolkit-Examples/ColorPicker/GazeableColorPicker.cs [hunk 6]
```diff
@@ -50,5 +36,30 @@ void UpdatePickedColor(PickedColorCallback cb)
...
+        {
+            gazing = true;
+        }
+
+        public void OnFocusExit()
+        {
+            gazing = false;
+        }
+
+        public void OnInputUp(InputEventData eventData)
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/ColorPicker/GazeableColorPicker.cs]: Sorry I should have thought of this in the interface review but just for handling a simple air tap every class with not have these two stub methods that do nothing. Thoughts?

Comment (maxouellet) [inline Assets/HoloToolkit-Examples/ColorPicker/GazeableColorPicker.cs]: Yeah, it might make sense. I agree that there's a bunch of cases where you only care about click, and other cases where you only care about up/down (and not click).

We could split the interface in two: IInputHandler (for up/down) and IInputClickHandler (for click). Would it make sense to do this as a separate pull request, or would you prefer I add it to this one?

File: Assets/HoloToolkit-Examples/SpatialUnderstanding/SpatialUnderstanding-FeatureOverview/Scripts/SpatialUnderstandingBasicCursor.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+using HoloToolkit.Unity.InputModule;
+using UnityEngine;
+
+/// <summary>
+/// 1. Decides when to show the cursor.
+/// 2. Positions the cursor at the gazed location.
+/// 3. Rotates the cursor to match hologram normals.
+/// </summary>
+public class SpatialUnderstandingBasicCursor : MonoBehaviour
```
Comment (maxouellet) [inline Assets/HoloToolkit-Examples/SpatialUnderstanding/SpatialUnderstanding-FeatureOverview/Scripts/SpatialUnderstandingBasicCursor.cs]: I can add it if you want, but it was never in there in the first place. Your call.
In any case, it would be HoloToolkit.Unity.Examples, or something like that?

File: Assets/HoloToolkit/Input/README.md [hunk 8]
```diff
@@ -1,6 +1,32 @@
...
+Each input source (hands, gestures, others) implement a **IInputSource** interface. The interface defines various events that the input sources can trigger. The input sources register themselves with the InputManager, whose role it is to forward input to the appropriate game objects. Input sources can be dynamically enabled / disabled as necessary, and new input sources can be created to support different input devices.
+
+Game objects that want to consume input events can implement one or many **input interfaces**, such as:
+
+- **IFocusable** for focus enter and exit. The focus can be triggered by the user's gaze or any other gaze source.
+- **IHoldHandle** for the Windows hold gesture.
+- **IInputHandler** for source up, down and clicked. The source can be a hand that tapped, a clicker that was pressed, etc.
+- **IManipulationHandler** for the Windows manipulation gesture.
+- **INavigationnHandler** for the Windows navigation gesture.
+- **ISourceStateHandler** for the source detected and source lost events.
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/README.md]: You don't need a SourceMoved event, because at any point in time, you can query the IInputSource (that you receive as part of the event data for any of the events) for its position and/or orientation, if available. HandDraggable gives a good example of that. This allows the event system to stay relatively lightweight in terms of number of events sent.

File: Assets/HoloToolkit/Input/README.md [hunk 9]
```diff
@@ -13,71 +39,106 @@ Torus shaped CursorOnHolograms when user is gazing at holograms and point light
...
-#### FocusedObjectKeywordManager.prefab
-Keyword manager pre-wired to send messages to object being currently focused via FocusedObjectMessageSender component.
-You can simply drop this into your scene and be able to send arbitrary messages to currently focused object.
+#### DefaultCursor.prefab
+3D animated cursor that follows the user's gaze and uses the Unity animation system to handle its various states. This cursor imitates the HoloLens Shell cursor.
 
-#### SelectedObjectKeywordManager.prefab
-Keyword manager pre-wired to send messages to object being currently selected via SelectedObjectMessageSender comoponent.
-You can simply drop this into your scene and be able to send arbitrary messages to currently selected object.
+#### HoloLensCamera.prefab
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/README.md]: It's much more powerful as far as customization goes, and it leverages the same scripts already used for simulating hand movement in the editor. This change remove the "Main Camera.prefab", as this ones fully replaces it.

File: Assets/HoloToolkit/Input/Scripts/BasicCursor.cs [hunk 10]
```diff
@@ -1,94 +0,0 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License. See LICENSE in the project root for license information.
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/BasicCursor.cs]: Because it isn't needed anymore. @paseb made a new ICursor interface + cursor scripts that are better designed, and include the functionalities tha the old cursor provided

File: Assets/HoloToolkit/Input/Scripts/Cursor/CursorModifier.cs [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+
+namespace HoloToolkit.Unity.InputModule
+{
+    /// <summary>
+    /// Component that can be added to any collider to modify how a cursor reacts when on that collider.
+    /// </summary>
+    public class CursorModifier : MonoBehaviour, ICursorModifier
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Cursor/CursorModifier.cs]: Yes, that's what the comment says. More specifically, it needs to be added to a game object with a collider

File: Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+                    GazeTransform = Camera.main.transform;
+                }
+                else
+                {
+                    Debug.LogError("Gaze Manager was not given a GazeTransform and no main camera exists to default to.");
+                }
+            }
+        }
+
+        private void Update()
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs]: I'm not setting it. I thought you already had a script that sets the focus point based on the gaze manager's output. Wouldn't it just be a matter of adding that script to a game object in the scene?

File: Assets/HoloToolkit/Input/Scripts/Cursor/AnimatedCursor.cs [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        /// <summary>
+        /// Data struct for cursor state information for the Animated Cursor
+        /// </summary>
+        [Serializable]
+        public struct AnimCursorDatum
+        {
+            public string Name;
+            public CursorStateEnum CursorState;
+
+            public enum AnimInputTypeEnum
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Cursor/AnimatedCursor.cs]: You mean the enum values for AnimInputTypeEnum? They just match the [Unity types for Unity animations](https://docs.unity3d.com/Manual/AnimationParameters.html).

I guess it might not be obvious for people that haven't used Unity's animations. I can add a comment on the enum if you want.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+            /// </summary>
+            Observe, 
+            /// <summary>
+            /// Not IsHandVisible AND not IsInputSourceDown AND TargetedObject exists
+            /// </summary>
+            ObserveHover, 
+            /// <summary>
+            /// IsHandVisible AND not IsInputSourceDown AND TargetedObject is NULL
+            /// </summary>
+            Interact,
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Yeah, it is. I think the way @paseb names the states, "Interact" means that the user's hand is visible with the finger in the ready state, so the user is in a "ready to interact." state. InteractHover is the equivalent for when the user's gaze is hovering over an object.

I don't object to renaming those if you have a better name for the states, but I don't have a better suggestion in mind.

File: Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+            if (EventSystem.current != null)
+            {
+                // NOTE: We need to do this AFTER we set the HitPosition and HitObject since we need to use HitPosition to perform the correct 2D UI Raycast.
+                RaycastUnityUI();
+            }
+
+            // Dispatch changed event if focus is different
+            if (previousFocusObject != HitObject && FocusedObjectChanged != null)
+            {
+                FocusedObjectChanged(previousFocusObject, HitObject);
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs]: No, it raises FocusedObjectChanged. You can get the OnFocusEnter and OnFocusExit message through the input module, by implementing the IFocusable interface in your MonoBehaviour

File: Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs (hunk 12 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs]: I agree with splitting the code in functions. From what I remember, it wasn't trivial to do because of dependencies between some variables, but can take a quick stab at it tomorrow.

I'll look at the focus point stuff as well. Might just be a matter of adding it to the InputManager prefab.

File: Assets/HoloToolkit-Examples/GazeRuler/Scenes/GazeRulerExample.unity [hunk 16]
```diff
@@ -349,71 +232,71 @@ MonoBehaviour:
...
             m_IntArgument: 0
             m_FloatArgument: 0
             m_StringArgument: 
             m_BoolArgument: 0
           m_CallState: 2
       m_TypeName: UnityEngine.Events.UnityEvent, UnityEngine, Version=0.0.0.0, Culture=neutral,
         PublicKeyToken=null
-  - Keyword: close
-    KeyCode: 0
+  - Keyword: 636c6f7365
```
Comment (maxouellet) [inline Assets/HoloToolkit-Examples/GazeRuler/Scenes/GazeRulerExample.unity]: Yeah that's weird. I'll take a look at the scene

Comment (HodgsonSDAS) [inline Assets/HoloToolkit-Examples/GazeRuler/Scenes/GazeRulerExample.unity]: Looks like the keyword is getting hashed?

File: Assets/HoloToolkit-Examples/SpatialUnderstanding/SpatialUnderstanding-FeatureOverview/Scripts/SpatialUnderstandingBasicCursor.cs (hunk 7 above)
Comment (HodgsonSDAS) [inline Assets/HoloToolkit-Examples/SpatialUnderstanding/SpatialUnderstanding-FeatureOverview/Scripts/SpatialUnderstandingBasicCursor.cs]: Yeah I'm not so sure example scripts should be in the HoloToolkit namespace unless they're apart of the Toolkit itself.

File: Assets/HoloToolkit/Input/Scripts/Cursor/CursorModifier.cs (hunk 11 above)
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Input/Scripts/Cursor/CursorModifier.cs]: We used the object's layer to do this as well. Not sure how straight forward that is though.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 14 above)
Comment (paseb) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Those map into the states any cursor could be in (hand visible / not visible and whether or not you're over a hologram) with the ability for a contextual override.  These should be part of the design learnings shared as part of the shell teams explorations.  That names are the closest approximation to what they reflect.  The comments on the enum states should be pretty straight forward.

File: Assets/HoloToolkit-Examples/GazeRuler/Scenes/GazeRulerExample.unity (hunk 16 above)
Comment (maxouellet) [inline Assets/HoloToolkit-Examples/GazeRuler/Scenes/GazeRulerExample.unity]: I don't know why this happened, but it was somehow corrupted. Looks like a hash gone wrong.
In any case, I fixed it, will be part of my next change..

File: Assets/HoloToolkit-Examples/SpatialUnderstanding/SpatialUnderstanding-FeatureOverview/Scripts/SpatialUnderstandingBasicCursor.cs (hunk 7 above)
Comment (maxouellet) [inline Assets/HoloToolkit-Examples/SpatialUnderstanding/SpatialUnderstanding-FeatureOverview/Scripts/SpatialUnderstandingBasicCursor.cs]: Looking at it, none of the scripts in the SpatialUnderstanding-FeatureOverview project are in a namespace. Given that, I'll leave SpatialUnderstandingBasicCursor in no namespace.

I agree that they should be changed so that each example is in its own namespace, but don't think it makes sense to do this work as part of the current pull request.

File: Assets/HoloToolkit/Input/README.md [hunk 17]
```diff
@@ -13,71 +39,106 @@ Torus shaped CursorOnHolograms when user is gazing at holograms and point light
...
+Interface that a game object can implement to react to manipulation gestures.
 
-**StabilityVarianceWeight** Stability variance weight multiplier factor.
+##### INavigationHandler.cs
+Interface that a game object can implement to react to source state changes, such as when an input source is detected or lost.
 
-#### GestureManager.cs
-GestureManager provides access to several different input gestures, including Tap and Manipulation. 
+##### ISourceStateHandler.cs
+Interface that a game object can implement to react to navigation gestures.
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/README.md]: Good catch, fixed

File: Assets/HoloToolkit/Input/README.md [hunk 18]
```diff
@@ -91,43 +152,65 @@ gesture is performed, all the relevant GameObjects will be moved.  If the desire
...
+##### EditorHandsInput.cs
+Input source for fake hands information, which can be used to simulate hands input in the Unity editor.
 
-**_HandGuidanceIndicator_** GameObject to display when your hand is about to lose tracking.
+##### GesturesInput.cs
+Input source for gestures information from the WSA APIs, which gives access to various system supported gestures.
 
-**HandGuidanceThreshold** When to start showing the HandGuidanceIndicator.  1 is out of view, 0 is centered in view.
+##### IInputSource.cs
+Interface for an input source. Defines
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/README.md]: Fixed, thanks!

File: Assets/HoloToolkit/Input/Scripts/Cursor/AnimatedCursor.cs (hunk 13 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Cursor/AnimatedCursor.cs]: Added some better comments + tooltip, that will be in my next push

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs [hunk 19]
```diff
@@ -0,0 +1,478 @@
...
+        }
+
+#endregion
+
+        /// <summary>
+        /// Register to events from the gaze manager, if not already registered.
+        /// </summary>
+        private void RegisterGazeManager()
+        {
+            if (!isRegisteredToGazeManager && gazeManager != null)
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Agreed, the state management here isn't needed anymore. I removed it, and merged registration / unregistration in two protected virtual methods called RegisterManagers and UnregisterManagers

File: Assets/HoloToolkit/Input/Scripts/Cursor/CursorModifier.cs (hunk 11 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Cursor/CursorModifier.cs]: I modified the comment a little to clarify it

File: Assets/HoloToolkit/Input/Scripts/Cursor/CursorModifier.cs [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        private void Awake()
```
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Cursor/CursorModifier.cs]: Agreed, done

File: Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs (hunk 12 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs]: Fixed the focus point by simplifying that StabilizationPlaneModifier script a little, and adding it to the InputManager prefab so that everyone uses it by default.

Also split the Update method of GazeManager to make it cleaner

File: Assets/HoloToolkit-Examples/ColorPicker/GazeableColorPicker.cs (hunk 6 above)
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/ColorPicker/GazeableColorPicker.cs]: I am OK if it's fixed in a separate one. Although, do you think the up/down might make sense in ISourceState?

File: Assets/HoloToolkit-Examples/SpatialUnderstanding/SpatialUnderstanding-FeatureOverview/Scripts/SpatialUnderstandingBasicCursor.cs (hunk 7 above)
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/SpatialUnderstanding/SpatialUnderstanding-FeatureOverview/Scripts/SpatialUnderstandingBasicCursor.cs]: Agreed. Please leave as-is.

File: Assets/HoloToolkit/Input/README.md (hunk 9 above)
Comment (NeerajW) [inline Assets/HoloToolkit/Input/README.md]: This would break the documentation links on dev center since they refer to the Main Camera.prefab in HTK. Perhaps file an issue to remind us to fix dev center with this change.

File: Assets/HoloToolkit-Examples/ColorPicker/GazeableColorPicker.cs (hunk 6 above)
Comment (maxouellet) [inline Assets/HoloToolkit-Examples/ColorPicker/GazeableColorPicker.cs]: It could, but ISourceStateHandler is already used for visibility. I think it'd make sense to do this:

ISourceStateHandler: source up/down
ISourceVisibilityHandler: source detected / lost

I can look at making that change tomorrow, it shouldn't be too bad.

**Comment by aalmada** [comment]
@maxouellet Is the InputManagerTest.scene gone? I liked that example with multiple forms of interaction.

**Comment by HodgsonSDAS** [comment]
It's still there @aalmada Assets\ThirdParty\HoloToolkit\Input\Tests\Scenes

**Comment by HodgsonSDAS** [comment]
@maxouellet After doing some tests in the InputManagerTest scene, I've found problems with the pop up menu on the cylinder we grow/shrink. Sometimes either the pop up menu doesn't go away, or it never pops in.  There's also some issues with the animations, but that's trivial.

**Comment by aalmada** [comment]
@HodgsonSDAS Sorry but I was misinterpreting the info for this PR. I was already looking in the local ExtensibleInputManager branch but it will be here only after the merge. Thanks.

File: Assets/HoloToolkit/Input/Scripts/InputManager.cs [hunk 21]
```diff
@@ -0,0 +1,581 @@
...
+        private void GazeManager_FocusedChanged(GameObject previousObject, GameObject newObject)
+        {
+            if (disabledRefCount > 0)
+            {
+                return;
+            }
+
+            if (previousObject != null)
+            {
+                ExecuteEvents.ExecuteHierarchy<IFocusable>(previousObject, null, (inputHandler, eventData) => { inputHandler.OnFocusExit(); });
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Input/Scripts/InputManager.cs]: Not so sure I like this line.  I wonder if there's a better way.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 14 above)
Comment (NeerajW) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: >>The names are the closest approximation to what they reflect. 
Not sure how true this is. Seems like it's the 'Ready' state per OOBE and other verbiage correct?

File: Assets/HoloToolkit/Input/Scripts/CursorManager.cs [hunk 22]
```diff
@@ -1,65 +0,0 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (NeerajW) [inline Assets/HoloToolkit/Input/Scripts/CursorManager.cs]: I want us to be careful with these statements. Let's have an inclusive mindset where we respect some people find value for these type of classes. I agree with the sentiment however that is a blanket argument. I will add it back if I see a need for it.

**Comment by maxouellet** [comment]
@HodgsonSDAS I have not been able to repro the cylinder popup issue you are mentioning.

What do you mean by disabling the debug? Do you mean the debug output for the events in the test scene? It's a test scene, meant to show concepts, so I think it's fine to have them in there. The debug log statements are in their own debug script, so it's not like they are polluting other scripts that you might want to reuse for your projects.

For the framerate on that scene, I think it's mostly because the content in that scene is using the standard shader. I can try to switch it to a different shader and see what it does. Again, not super concerned about it, since it's a test scene that we can improve as needed.

File: Assets/HoloToolkit/Input/Scripts/InputManager.cs (hunk 21 above)
Comment (maxouellet) [inline Assets/HoloToolkit/Input/Scripts/InputManager.cs]: Agreed that this might save a few allocations. I'd suggest making that optimization at a later time, should it become necessary. It wouldn't be a breaking change, since it's an internal implementation detail of InputManager, and I'd like to get this first version in.

Let me know if you disagree.

**Comment by HodgsonSDAS** [comment]
When building on device the debug scripts really impact performance.  I understand it's a test scene, and debug logs are good for showing off exactly what's happening, but I think having a way to disable it might be nice.

**Comment by HodgsonSDAS** [comment]
I'll have to get back to you on the repo for the cylinder.

File: Assets/HoloToolkit/Input/Scripts/InputManager.cs (hunk 21 above)
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Input/Scripts/InputManager.cs]: No worries, I wouldn't mind doing this work if needed later on.

**Comment by maxouellet** [comment]
My guess is that the major debug log perf impact comes from the NavigationUpdated and ManipulationUpdated events being broadcasted many times a second when you're dragging an object. I can disable those behind a boolean (and default them to disabled), I think that'd help.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 14 above)
Comment (paseb) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: These are what we used to describe the states.  As long as the comments correctly reflect what the state indicates the word you use for it is purely semantics.  'Ready' is a nebulous term from a cursor and input standpoint IMO.

**Comment by HodgsonSDAS** [comment]
@maxouellet Repro steps for Cylinder:

Start scene
Open modal input popup in center panel
click on cylinder, no pop up.

**Comment by maxouellet** [comment]
@HodgsonSDAS If I understand your repro steps well, that's by design. When you open the modal input popup, it disabled all other input until the popup is dismissed. So until you click the "Exit" button that is in the popup, all other input won't do anything.

**Comment by HodgsonSDAS** [comment]
Ohh, wow, I can't read.

**Comment by HodgsonSDAS** [comment]
Okay, one last thing. If you walk through the scene objects the rendering does some crazy stuff until you look at the scene objects again.

File: Assets/HoloToolkit/Input/Scripts/CursorManager.cs (hunk 22 above)
Comment (paseb) [inline Assets/HoloToolkit/Input/Scripts/CursorManager.cs]: I want us to be honest.  Keeping deprecated classes can cause confusion and creates a fragmented code base.  We have to move others forward with solid examples that can be built off of rather than having sentimental attachment to prior iterations.  If anyone wants to get older versions they are free to do so.

File: Assets/HoloToolkit-Examples/ColorPicker/GazeableColorPicker.cs (hunk 6 above)
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/ColorPicker/GazeableColorPicker.cs]: Sweet!

File: Assets/HoloToolkit/Input/Scripts/CursorManager.cs (hunk 22 above)
Comment (NeerajW) [inline Assets/HoloToolkit/Input/Scripts/CursorManager.cs]: It is not sentimental. It is ensuring you are actually helping developers along the way.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 14 above)
Comment (NeerajW) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Feedback is a gift :) 
If Ready is nebulous then the OOBE experience should not be using that verbiage.
Anyway I can see pushback on this. So please consider this non-blocking.

**Comment by NeerajW** [comment]
Looks like only blocking comment left is what HodgonSDAS reported.

>>If you walk through the scene objects the rendering of the cursor does some crazy stuff until you look at the scene objects again. 

This would defeat the purpose of all this :)

**Comment by maxouellet** [comment]
@HodgsonSDAS Just tried this on device. Do you mean that if you approach an object that has a collider, your cursor will eventually disappear if that object gets closer than 0.85 meters (current near clip plane)? And then the cursor reappears as soon as you target an object farther than the near clip plane.

If this is what you mean, then this is not a new issue: it already existed with the previous cursor + gaze manager code (just validated by testing with the current HoloToolkit). Given that it is not a new issue, I'd be tempted to leave it as-is for now and fix it as a separate change if needed. I don't know that we really want to make the cursor go through objects that are in the near clip plane: there might be scenarios where that is not desirable.

If this is not what you meant, would love to have more details about this.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 14 above)
Comment (paseb) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: I polled a couple others and they also said "Ready" would be good to swap out for "Interact" as well.  My comment was toward other input devices not just gesture and what term is more applicable to cursor and input UX design.  What does "Ready" mean when using 3DoF input from the clicker?  Food for thought.  Good feedback, anything that makes more sense from broader perspective I'm all for.

**Comment by NeerajW** [comment]
Agreed with @maxouellet.  I understood something else.

File: Assets/HoloToolkit/Input/Scripts/InputManager.cs (hunk 21 above)
Comment (aalmada) [inline Assets/HoloToolkit/Input/Scripts/InputManager.cs]: Is the issue with the use of a lambda? The alternative is to have the code spread in multiple places. Look how the InputManager looks because of this.

**Comment by HodgsonSDAS** [comment]
No this isn't about the cursor clipping the near plane, it's a rendering issue where all the colors get separated like the GPU is having a hard time rendering everything. I see the red green and blue colors separately, and the sequential colors get rendered very far apart.

**Comment by paseb** [comment]
@HodgsonSDAS I believe this is a known color separation issue with the device when having a full white cursor.  This is compounded when framerate drops and interpolation speed is minimal.  A simple test to mitigate this is to shift the cursor color from being full white to #7BCEFFFF or a slight grey #C8C8C8FF.  I've also noticed perf issues with the default textmesh shader and backface transparency.  You should see it if you go to the back side of the panels and look back to see the 2 panels text overlapping from the reverse side.  There has been some additional research at methods to mitigate further the color separation artifacts using some rendering tricks that I can expand on if need be.

Also noted at the bottom "Color Separation"
https://developer.microsoft.com/en-us/windows/holographic/hologram_stability

**Comment by HodgsonSDAS** [comment]
@paseb I can understand if I'm moving my head quickly but this was happening even when standing still.

**Comment by maxouellet** [comment]
Ok, with your repro step, I've repro'ed the issue on my side. Will take a look, this is definitely weird

**Comment by paseb** [comment]
@HodgsonSDAS I was able to repro as well.  @maxouellet it looks like an issue with the GazeStabilityPlane when you go past the panels.  The plane remains exactly where the camera is and so does the focus point.

**Comment by HodgsonSDAS** [comment]
https://www.youtube.com/watch?v=G6QtXjueFms Pause at 0.07 seconds

**Comment by HodgsonSDAS** [comment]
@paseb Yeah, I think you're right.

**Comment by paseb** [comment]
override

Yep for a quick fix or to validate is to use the Default cursor as the override for the StabilizationPlaneModifier transform to test.  @maxouellet is working on a better fix for the issue.

**Comment by maxouellet** [comment]
Thanks a lot for finding this @HodgsonSDAS , and thanks @paseb for the help. Pushed a change that fixes the issue by using the default plane distance when no object is hit, as opposed to using the last hit distance. Also fixed a bug in StabilizationPlaneModifier that made it not respect the "Draw Gizmos" option: it now behaves properly.

**Comment by HodgsonSDAS** [comment]
@maxouellet , @paseb  you guys are awesome. Sorry for being such a pain in the ass. I just want this stuff to be the best it can be. haha.

**Comment by HodgsonSDAS** [comment]
Before we merge this PR, could be merge Microsoft:Master into Microsoft:ExtensibleInputModule?

**Comment by NeerajW** [comment]
@HodgsonSDAS thank you for being so diligent. Rock on!
